_posts/2017/2017-10-30-typescript-codesplitting-treeshaking.md

```diff
@@ -0,0 +1,111 @@
+    "jsx": "preserve",
+    // resolve modules as you would expect
+    "moduleResolution": "node",
+    // leave imports as they are
+    "module": "esnext",
+    // do not transpile stuff like classes, async/await, ...
+    "target": "esnext",
+    // produce a source map
+    "sourceMap": true
+  }
+}
+{% endhighlight %}
+
+## Babel configuration
+
+Now you will get JavaScript that uses a bunch of features, that are not widely supported by now. With the preset `babel-preset-env` you can define not what exactly [Babel][babel] transform you want to apply, but what browsers you want to support and then `babel-preset-env` picks the right configuration to match your needs. Your `.babelrc` should look somewhat like this:
+
+{% highlight javascript %}
+{
+  "presets": [
+    "react",
+    [
+      "env",
+      {
+        // leave imports as they are
+        "modules": false,
+        "targets": {
+          "browsers": [
+            // choose browsers you want to support
+            "last 2 chrome versions"
+          ]
+        }
+      }
+    ]
+  ],
+  "plugins": [
+    // support dynamic import syntax, but leave it unchanged
+    "babel-plugin-syntax-dynamic-import"
+  ]
+}
+{% endhighlight %}
+
+Two important things to notice here: You must tell `babel-preset-env` to not touch the modules, meaning it will just keep import statements like `import a from './b'` as they are. In addition you have to add `babel-plugin-syntax-dynamic-import` as plugin to keep dynamic import statements like `const a = await import('./b')` untouched. After this step, you now have compiled your TypeScript down to JavaScript that is compatible with the browsers of your choice, but with the original (dynamic) imports.
+
+## Webpack configuration
+
+As final step, we have to configure Webpack accordingly. You would use your `awesome-typescript-loader` as usual, but add the `useBabel: true` option (which is the same as prepending the `babel-loader` in the rules chain). The important parts of your `webpack.config.js` would look like this:
+
+{% highlight javascript %}
+module.exports = {
+  target: 'web',
+  entry: {
+    app: ['./index.ts'],
+  },
+  module: {
+    rules: [
+      {
+        test: /\.tsx?$/,
+        use: [
+          {
+            loader: require.resolve('awesome-typescript-loader'),
+            options: {
+              // compile with TypeScript, then transpile with Babel
+              useBabel: true,
+            },
+          },
+        ],
+      },
+    ],
+  },
+  // produce sourcemap
+  devtool: '#sourcemap',
+}
+{% endhighlight %}
+
+That is it. Now you not only enabled Webpack to do code splitting and tree shaking, but you also reduced your JavaScript asset size even more, because the use of `babel-preset-env` instead of letting TypeScript do the whole compilation. For example `async`/`await` will most likely just be in your resulting JavaScript instead of being transpiled to generators, which is just way shorter.
```

Comment:
what is meant by "transpiled"? Couldn't find this...

Reply:
https://www.stevefenton.co.uk/2012/11/compiling-vs-transpiling/